src/convert.rs

```diff
@@ -424,132 +406,111 @@ pub fn elf_to_tbf(
-    // elf2tab needs to be re-written to use segments rather the hack we have
-    // here. We just assume we can ad-hoc determine the sections we need to
-    // include, when we should just use the segment mapping. Because of this,
-    // `last_section_address_end` is a hack as well.
-    let mut last_section_address_end: Option<usize> = None;
-
-    // Iterate the sections in the ELF file. The sections are sorted in order of
-    // offset. Add the sections we need to the binary.
-    for s in &sections_sort {
-        let section = &input.sections[s.0];
+    let mut last_segment_address_end: Option<usize> = None;
 
-        // If this is writeable, executable, or allocated, is nonzero length,
-        // and is type `PROGBITS` we want to add it to the binary.
-        if (section.shdr.flags.0
-            & (elf::types::SHF_WRITE.0 + elf::types::SHF_EXECINSTR.0 + elf::types::SHF_ALLOC.0)
-            != 0)
-            && section.shdr.shtype == elf::types::SHT_PROGBITS
-            && section.shdr.size > 0
-        {
-            // This is a section we are going to add to the binary.
-
-            if last_section_address_end.is_some() {
-                // We have a previous section. Now, check if there is any
-                // padding between the sections in the .elf.
-                let end = last_section_address_end.unwrap();
-                let start = section.shdr.addr as usize;
-
-                // Because we have flash and ram memory regions, we have
-                // multiple address spaces. This check lets us assume we are in
-                // a new address segment. We need the start of the next section
-                // to be after the previous one, and the gap to not be _too_
-                // large.
-                if start > end {
-                    // If this is the next section in the same segment, then
-                    // check if there is any padding required.
-                    let padding = start - end;
-
-                    if padding < 1024 {
-                        if padding > 0 {
-                            if verbose {
-                                println!("  Adding {} bytes of padding between sections", padding,);
+    let mut start_address: u64 = 0;
+    let start_program_binary_index = binary_index;
+
+    // Iterate over ELF's Program Headers to assemble the binary image as a contiguous
+    // memory block. Only take into consideration segments where filesz is greater than 0
+    for segment in &elf_file.phdrs {
+        match segment.progtype {
+            elf::types::PT_LOAD => {
+                if segment.filesz > 0 {
+                    if last_segment_address_end.is_some() {
+                        // We have a previous section. Now, check if there is any
+                        // padding between the sections in the .elf.
+                        let padding = segment.paddr as usize - last_segment_address_end.unwrap();
+                        if padding < 1024 {
```

Comment:
We shouldn't need this check. Since we using the segments, if the elf tells us to include the segment then we should just do it and insert padding as needed. We don't have to worry about using a section we shouldn't be using like we did before.